File: python_src/13.罗马数字转整数.py
#
# @lc app=leetcode.cn id=13 lang=python3
#
# [13] 罗马数字转整数
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        d = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        num = 0
        last = 0
        for c in reversed(s):
            num += d[c] * (1 if d[c] >= last else -1)
            last = max(d[c], last)
        return num
# @lc code=end

s = Solution()
print(s.romanToInt('III'))
print(s.romanToInt('IV'))
print(s.romanToInt('IX'))
print(s.romanToInt('LVIII'))
print(s.romanToInt('MCMXCIV'))

File: rust_src/13.罗马数字转整数.rs
/*
 * @lc app=leetcode.cn id=13 lang=rust
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
		let mut res = 0;
		let mut last = ' ';

		for c in s.chars().rev()
		{
			match c
			{
				'I' =>
				{
					if last == 'V' || last == 'X'
					{
						res -= 1;
					}
					else
					{
						res += 1;
					}
				}
				'X' =>
				{
					if last == 'L' || last == 'C'
					{
						res -= 10;
					}
					else
					{
						res += 10;
					}
				}
				'C' =>
				{
					if last == 'D' || last == 'M'
					{
						res -= 100;
					}
					else
					{
						res += 100;
					}
				}
				'V' => res += 5,
				'L' => res += 50,
				'D' => res += 500,
				'M' => res += 1000,
				_ => (),
			}

			last = c;
		}

		return res;
    }
}
// @lc code=end

